core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,392 @@
+ * The data transfer is done using object storage low-level multipart upload.
+ * <p>
+ * We upload data in partitions. When write(), the data will be persisted to
+ * a temporary file {@link #mFile} on the local disk. When the data {@link #mPartitionOffset}
+ * in this temporary file reaches the {@link #mPartitionSize}, the file will be submitted
+ * to the upload executor {@link #mExecutor} and we do not wait for uploads to finish.
+ * A new temp file will be created for the future write and the {@link #mPartitionOffset}
+ * will be reset to zero. The process goes until all the data has been written to temp files.
+ * <p>
+ * In flush(), we upload the buffered data if they are bigger than 5MB
+ * and wait for all uploads to finish. The temp files will be deleted after uploading successfully.
+ * <p>
+ * In close(), we upload the last part of data (if exists), wait for all uploads to finish,
+ * and complete the multipart upload.
+ * <p>
+ * close() will not be retried, but all the multipart upload
+ * related operations(init, upload, complete, and abort) will be retried.
+ * <p>
+ * If an error occurs and we have no way to recover, we abort the multipart uploads.
+ * Some multipart uploads may not be completed/aborted in normal ways and need periodical cleanup
+ * by enabling the {@link PropertyKey#UNDERFS_CLEANUP_ENABLED}.
+ * When a leader master starts or a cleanup interval is reached, all the multipart uploads
+ * older than clean age will be cleaned.
+ */
+@NotThreadSafe
+public abstract class ObjectLowLevelOutputStream extends OutputStream {
+  protected static final Logger LOG = LoggerFactory.getLogger(ObjectLowLevelOutputStream.class);
+
+  protected final List<String> mTmpDirs;
+
+  /**
+   * Only parts bigger than 5MB could be uploaded through multipart upload,
+   * except the last part.
+   */
+  protected static final long UPLOAD_THRESHOLD = 5L * Constants.MB;
+
+  /** Bucket name of the object storage bucket. */
+  protected final String mBucketName;
+
+  /** Key of the file when it is uploaded to object storage. */
+  protected final String mKey;
+
+  /** The retry policy of this multipart upload. */
+  protected final RetryPolicy mRetryPolicy = new CountingRetry(5);
+
+  /** Pre-allocated byte buffer for writing single characters. */
+  protected final byte[] mSingleCharWrite = new byte[1];
+
+  /** The MD5 hash of the file. */
+  protected MessageDigest mHash;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  protected boolean mClosed = false;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  protected long mPartitionOffset;
+  /** The maximum allowed size of a partition. */
+  protected final long mPartitionSize;
+
+  /**
+   * The local temp file that will be uploaded when reaches the partition size
+   * or when flush() is called and this file is bigger than {@link #UPLOAD_THRESHOLD}.
+   */
+  protected File mFile;
```

Comment:
can use optional<File> and Optional<OutputStream> for variables that can have null values